flink-table/flink-table-planner/src/main/java/org/apache/calcite/sql/fun/SqlCastFunction.java

```diff
@@ -98,7 +103,21 @@ public RelDataType inferReturnType(SqlOperatorBinding opBinding) {
         assert opBinding.getOperandCount() == 2;
         RelDataType ret = opBinding.getOperandType(1);
         RelDataType firstType = opBinding.getOperandType(0);
-        ret = opBinding.getTypeFactory().createTypeWithNullability(ret, firstType.isNullable());
+
+        RelOptCluster relOptCluster =
+                ((FlinkCalciteSqlValidator) ((SqlCallBinding) opBinding).getValidator())
```

Comment:
Can you check if opBinding is an instance of SqlCallBinding just like the if block below?

What about ?
```
        boolean forceOutputTypeNullable = false;
        if (opBinding instanceof SqlCallBinding
                && ((SqlCallBinding) opBinding).getValidator()
                        instanceof FlinkCalciteSqlValidator) {
            FlinkCalciteSqlValidator validator =
                    (FlinkCalciteSqlValidator) ((SqlCallBinding) opBinding).getValidator();

            TableConfig tableConfig =
                    ShortcutUtils.unwrapContext(validator.getRelOptCluster()).getTableConfig();
            forceOutputTypeNullable =
                    tableConfig
                            .get(ExecutionConfigOptions.TABLE_EXEC_LEGACY_CAST_BEHAVIOUR)
                            .isEnabled();
        }

        ret =
                opBinding
                        .getTypeFactory()
                        .createTypeWithNullability(
                                ret, firstType.isNullable() || forceOutputTypeNullable);
```

Reply:
fixed